File: src/ReCrafted/Graphics/Camera.cs
// ReCrafted © 2016 Damian 'Erdroy' Korczowski

using System;
using System.Windows.Forms;
using ReCrafted.Core;
using SharpDX;

namespace ReCrafted.Graphics
{
    /// <summary>
    /// Camera class.
    /// </summary>
    public sealed class Camera : ITickable
    {
        /// <summary>
        /// The current camera which is used to render everything.
        /// </summary>
        public static Camera Current;

        private readonly Vector3[] _filtering = new Vector3[3];
        private int _filteringStep;

        private float _aspectRatio;
        private Vector3 _rotation;
        private Vector3 _lastDelta;

        /// <summary>
        /// Camera contstructor
        /// </summary>
        public Camera()
        {
            BackgroundColor = new Color(0.3f, 0.3f, 0.3f, 1.0f);

            NearZPlane = 0.02f;
            FarZPlane = 1000.0f;
            FieldOfView = 0.75f;

            LookAcceleration = true;
            LookFiltering = true;
            UseLook = true;

            Forward = Vector3.ForwardLH;
        }

        /// <summary>
        /// Sets this camera as `current` for rendering.
        /// </summary>
        public void SetAsCurrent()
        {
            Current = this;

            if (FreeLook)
            {
                // zeroe the cursor position
                var clientCenter = Game.Instance.Form.PointToScreen(new System.Drawing.Point(Game.Instance.Form.ClientSize.Width / 2, Game.Instance.Form.ClientSize.Height / 2));
                Cursor.Position = clientCenter;
            }
        }

        /// <summary>
        /// Ticks the camera, this is called by the game loop, 
        /// no need to call it again manually.
        /// </summary>
        public void Tick()
        {
            // check if this camera is active
            if (IsActive)
                return;

            if (FreeLook)
            {
                // update
                UpdateMovement();

                if (UseLook && Input.IsButton(ButtonCode.Right))
                {
                    UpdateLook();

                    Cursor.Position = Game.Instance.Form.PointToScreen(
                        new System.Drawing.Point(Game.Instance.Form.ClientSize.Width/2,
                            Game.Instance.Form.ClientSize.Height/2));
                    Cursor.Hide();
                }
                else
                {
                    Cursor.Show();
                }
            }

            UpdateMatrices();
        }

        /// <summary>
        /// Updates the view/projection matrices.
        /// </summary>
        public void UpdateMatrices()
        {
            _aspectRatio = Display.ClientWidth / (float)Display.ClientHeight;

            ViewMatrix = Matrix.LookAtLH(Position, Position + Forward, Vector3.Up);
            ProjectionMatrix = Matrix.PerspectiveFovLH(FieldOfView, _aspectRatio, NearZPlane, FarZPlane);
            
            CameraFrustum = new BoundingFrustum(ViewProjectionMatrix);
        }

        /// <summary>
        /// Calculates rotation matrix.
        /// </summary>
        /// <returns>The rotation matrix.</returns>
        public Matrix GetRotationMatrix()
        {
            var rot = CreateFromYawPitchRoll(MathUtil.DegreesToRadians(-_rotation.Y), MathUtil.DegreesToRadians(-_rotation.X), 0.0f);
            var rotMatrix = Matrix.RotationQuaternion(rot);

            return rotMatrix;
        }

        // private
        private void UpdateMovement()
        {
            var rotMatrix = GetRotationMatrix();

            var dir = Vector3.Zero;

            if (Input.IsKey(KeyCode.W))
                dir += rotMatrix.Forward;

            if (Input.IsKey(KeyCode.S))
                dir -= rotMatrix.Forward;

            if (Input.IsKey(KeyCode.A))
                dir += rotMatrix.Left;

            if (Input.IsKey(KeyCode.D))
                dir -= rotMatrix.Left;

            if (Input.IsKey(KeyCode.Q))
                dir += rotMatrix.Up;

            if (Input.IsKey(KeyCode.E))
                dir -= rotMatrix.Up;

            dir.Normalize();

            // speed
            if (Input.IsKey(KeyCode.ShiftKey))
                dir *= 15.0f;
            else if (Input.IsKey(KeyCode.ControlKey))
                dir *= 1.0f;
            else
                dir *= 6.0f;

            Position += dir * Time.DeltaTime;
        }

        // private
        private void UpdateLook()
        {
            var cursorPos = Cursor.Position;
            var clientCenter = Game.Instance.Form.PointToScreen(new System.Drawing.Point(Game.Instance.Form.ClientSize.Width / 2, Game.Instance.Form.ClientSize.Height / 2));
            var delta = new System.Drawing.Point(cursorPos.X - clientCenter.X, cursorPos.Y - clientCenter.Y);

            var mouseDelta = new Vector3(delta.X, delta.Y, 0.0f);

            if (LookFiltering)
            {
                // Filtering
                _filtering[_filteringStep] = new Vector3(delta.X, delta.Y, 0.0f);
                _filteringStep++;

                if (_filteringStep == 3)
                    _filteringStep = 0;

                // Calculate avg. mouse delta
                mouseDelta = (_filtering[0] + _filtering[1] + _filtering[2]) / new Vector3(3.0f, 3.0f, 1.0f);
            }

            // acceleration
            var accel = mouseDelta;

            if (LookAcceleration)
                accel += _lastDelta;

            _rotation += new Vector3(accel.X / 16.0f, accel.Y / 16.0f, 0.0f);
            _rotation.Y = MathUtil.Clamp(_rotation.Y, -89.0f, 89.0f);

            var rotMatrix = GetRotationMatrix();
            Forward = rotMatrix.Forward;

            _lastDelta = mouseDelta;

            Cursor.Position = clientCenter;
        }

        // private
        private static Quaternion CreateFromYawPitchRoll(float yaw, float pitch, float roll)
        {
            var rollOver2 = roll * 0.5f;
            var sinRollOver2 = (float)Math.Sin(rollOver2);
            var cosRollOver2 = (float)Math.Cos(rollOver2);
            var pitchOver2 = pitch * 0.5f;
            var sinPitchOver2 = (float)Math.Sin(pitchOver2);
            var cosPitchOver2 = (float)Math.Cos(pitchOver2);
            var yawOver2 = yaw * 0.5f;
            var sinYawOver2 = (float)Math.Sin(yawOver2);
            var cosYawOver2 = (float)Math.Cos(yawOver2);

            Quaternion result;
            result.X = cosYawOver2 * cosPitchOver2 * cosRollOver2 + sinYawOver2 * sinPitchOver2 * sinRollOver2;
            result.Y = cosYawOver2 * cosPitchOver2 * sinRollOver2 - sinYawOver2 * sinPitchOver2 * cosRollOver2;
            result.Z = cosYawOver2 * sinPitchOver2 * cosRollOver2 + sinYawOver2 * cosPitchOver2 * sinRollOver2;
            result.W = sinYawOver2 * cosPitchOver2 * cosRollOver2 - cosYawOver2 * sinPitchOver2 * sinRollOver2;
            return result;
        }

        // properties
        // camera info
        /// <summary>
        /// The background color, which will be used 
        /// as `clearing` color for render targets.
        /// </summary>
        public Color BackgroundColor { get; set; }

        /// <summary>
        /// Make the look acceleration enabled or disabled.
        /// </summary>
        public bool LookAcceleration { get; set; }

        /// <summary>
        /// Make the look filtering enabled or disabled.
        /// </summary>
        public bool LookFiltering { get; set; }

        /// <summary>
        /// Can camera use look at the moment?
        /// Default: true.
        /// </summary>
        public bool UseLook { get; set; }

        /// <summary>
        /// The far z plane.
        /// </summary>
        public float FarZPlane { get; set; }

        /// <summary>
        /// The near z plane.
        /// </summary>
        public float NearZPlane { get; set; }

        /// <summary>
        /// The field of view.
        /// </summary>
        public float FieldOfView { get; set; }

        /// <summary>
        /// Let this camera have a free look?
        /// </summary>
        public bool FreeLook { get; set; }

        // transform
        /// <summary>
        /// Camera position.
        /// </summary>
        public Vector3 Position { get; set; }

        /// <summary>
        /// The look/forward direction.
        /// </summary>
        public Vector3 Forward { get; set; }
        
        /// <summary>
        /// The down direction.
        /// </summary>
        public Vector3 Down => ViewMatrix.Down;

        /// <summary>
        /// The backward direction.
        /// </summary>
        public Vector3 Backward => ViewMatrix.Backward;

        /// <summary>
        /// The left direction.
        /// </summary>
        public Vector3 Left => ViewMatrix.Left;

        /// <summary>
        /// The right direction.
        /// </summary>
        public Vector3 Right => ViewMatrix.Right;

        /// <summary>
        /// True when this camera is the
        /// current rendering target camera.
        /// </summary>
        public bool IsActive => Current != this;

        /// <summary>
        /// Camera's frustum.
        /// </summary>
        public BoundingFrustum CameraFrustum { get; private set; }

        // matrices
        /// <summary>
        /// The View matrix.
        /// </summary>
        public Matrix ViewMatrix { get; private set; }

        /// <summary>
        /// The Projection matrix.
        /// </summary>
        public Matrix ProjectionMatrix { get; private set; }

        /// <summary>
        /// The View*Projection matrix.
        /// </summary>
        public Matrix ViewProjectionMatrix => ViewMatrix*ProjectionMatrix;
    }
}

File: src/ReCrafted/Graphics/Camera.cpp
// ReCrafted © 2016-2017 Damian 'Erdroy' Korczowski and Mateusz 'Maturas' Zawistowski

#include "Camera.h"
#include "../Core/GameCore.h"

// declare m_mainCamera static variable
Camera* Camera::m_mainCamera;

// https://github.com/Erdroy/ReCrafted/blob/9138a5067abb88ddd3fc5dd2059c1d72e2b67535/src/ReCrafted/Graphics/Camera.cs

void Camera::updateRotation()
{
	auto yaw = Math::degreeToRadian(-m_rotation.y);
	auto pitch = Math::degreeToRadian(-m_rotation.x);
	auto roll = 0.0f;

	auto yawOver2 = yaw * 0.5;
	auto cosYawOver2 = float(cos(yawOver2));
	auto sinYawOver2 = float(sin(yawOver2));

	auto pitchOver2 = pitch * 0.5;
	auto cosPitchOver2 = float(cos(pitchOver2));
	auto sinPitchOver2 = float(sin(pitchOver2));

	auto rollOver2 = roll * 0.5;
	auto cosRollOver2 = float(cos(rollOver2));
	auto sinRollOver2 = float(sin(rollOver2));
	
	// quaternion - from euler
	auto W = cosYawOver2 * cosPitchOver2 * cosRollOver2 + sinYawOver2 * sinPitchOver2 * sinRollOver2;
	auto X = sinYawOver2 * cosPitchOver2 * cosRollOver2 + cosYawOver2 * sinPitchOver2 * sinRollOver2;
	auto Y = cosYawOver2 * sinPitchOver2 * cosRollOver2 - sinYawOver2 * cosPitchOver2 * sinRollOver2;
	auto Z = cosYawOver2 * cosPitchOver2 * sinRollOver2 - sinYawOver2 * sinPitchOver2 * cosRollOver2;

	// quaternion * axis
	auto single = X * 2.0f;
	auto single1 = Y * 2.0f;
	auto single2 = Z * 2.0f;
	auto single3 = X * single;
	auto single4 = Y * single1;
	auto single5 = Z * single2;
	auto single6 = X * single1;
	auto single7 = X * single2;
	auto single8 = Y * single2;
	auto single9 = W * single;
	auto single10 = W * single1;
	auto single11 = W * single2;

	// calculate forward
	auto axis = Vector3(0.0f, 0.0f, 1.0f);
	m_forward = Vector3(
		(1.0f - (single4 + single5)) * axis.x + (single6 - single11) * axis.y + (single7 + single10) * axis.z,
		(single6 + single11) * axis.x + (1.0f - (single3 + single5)) * axis.y + (single8 - single9) * axis.z,
		(single7 - single10) * axis.x + (single8 + single9) * axis.y + (1.0f - (single3 + single4)) * axis.z
	);

	// calculate right
	axis = Vector3(1.0f, 0.0f, 0.0f);
	m_right = Vector3(
		(1.0f - (single4 + single5)) * axis.x + (single6 - single11) * axis.y + (single7 + single10) * axis.z,
		(single6 + single11) * axis.x + (1.0f - (single3 + single5)) * axis.y + (single8 - single9) * axis.z,
		(single7 - single10) * axis.x + (single8 + single9) * axis.y + (1.0f - (single3 + single4)) * axis.z
	);

	// calculate up
	m_up = Vector3::cross(m_forward, m_right);
}

void Camera::updateControls()
{
	// update movement
	Vector3 direction = {};

	if (Input::isKey(Key_W))
		direction += m_forward;

	if (Input::isKey(Key_S))
		direction -= m_forward;

	if (Input::isKey(Key_A))
		direction -= m_right;

	if (Input::isKey(Key_D))
		direction += m_right;

	if (Input::isKey(Key_Q))
		direction -= m_up;

	if (Input::isKey(Key_E))
		direction += m_up;

	direction.normalize();

	direction *= float(Time::deltaTime());

	// shift to speed up
	auto mul = Input::isKey(Key_Shift) ? 50.0f : 15.0f;

	// ctrl to slow down
	mul = Input::isKey(Key_Control) ? 1.0f : mul;

	direction *= mul;

	if(direction.length() > 0.0f)
		m_position += direction;

	// update look
	auto center = Vector2(
		round(GameCore::getWindowWidth() / 2.0f), 
		round(GameCore::getWindowHeight() / 2.0f)
	);

	auto cursorPos = Input::getCursorPos();
	auto rawDelta = center - cursorPos;

	// update filtering buffer
	m_cursorDeltaBuffer[m_cursorDeltaBufferPosition] = rawDelta;
	m_cursorDeltaBufferPosition++;

	// reset if needed(out of bounds)
	if (m_cursorDeltaBufferPosition >= FILTERING_BUFFER_SIZE)
		m_cursorDeltaBufferPosition = 0u;

	// calculate delta
	Vector2 delta = {};
	for (auto i = 0u; i < FILTERING_BUFFER_SIZE; i++)
	{
		delta += m_cursorDeltaBuffer[i];
	}

	delta.x = delta.x / FILTERING_BUFFER_SIZE;
	delta.y = delta.y / FILTERING_BUFFER_SIZE;

	// accelerate
	auto accelDelta = delta + m_lastDelta;

	// apply camera rotation
	m_rotation += Vector3(accelDelta.x / 16.0f, accelDelta.y / 16.0f, 0.0f);
	m_rotation.y = Math::clamp(m_rotation.y, -89.9f, 89.9f);

	// update camera rotation
	updateRotation();

	// save needed data
	m_lastDelta = delta;
}

void Camera::updatePerspective()
{
	// create projection matrix
	Matrix::createPerspectiveFovLH(Math::degreeToRadian(m_fov), GameCore::getAspectRatio(), m_nearPlane, m_farPlane, &m_projection);
}

void Camera::update()
{
	// update movement if this camera is 'free'.
	if (m_freeMovement && Input::isKey(Key_Mouse1)) 
	{
		updateControls(); // just do it!
		
		// lock position
		Input::setCursorPos(
			static_cast<int>(round(GameCore::getWindowWidth() / 2.0f)), 
			static_cast<int>(round(GameCore::getWindowHeight() / 2.0f))
		);
	}

	// update matrices
	m_lookAt = m_position + m_forward;

	// create view matrix
	Matrix::createViewLH(m_position, m_lookAt, m_upLock, &m_view);

	// update frustum
	// TODO: camera frustum for culling
}
